soccer/gameplay/skills/capture.py

```diff
@@ -1,270 +1,302 @@
+    PROBABLY_HELD_HISTORY_LENGTH = 60
 
-    # Time in which to wait in delay state to confirm the robot has the ball
-    DelayTime = 0.5
+    PROBABLY_HELD_START = .5 * PROBABLY_HELD_HISTORY_LENGTH
 
-    # Default dribbler speed, can be overriden by self.dribbler_power
-    # Sets dribbler speed during intercept and fine approach
-    DribbleSpeed = 100
+    # Anything below this number we think we don't have the ball
+    PROBABLY_NOT_HELD_CUTOFF = .4 * PROBABLY_HELD_HISTORY_LENGTH
+    # Anything above this number we think we have the ball
+    PROBABLY_HELD_CUTOFF = .6 * PROBABLY_HELD_HISTORY_LENGTH
 
-    # The minimum dot product result between the ball and the robot to count as the ball moving at the
-    # robot
-    InFrontOfBallCosOfAngleThreshold = 0.95
 
-    DelaySpeed = 0.2
+    # Accounts for the scaling needed to balance the robot change cost
+    # Increase to decrease the chance of robot change
+    # Decrease to improve the responsiveness in regard th changing robots
+    SETTLE_COST_MULTIPLIER = 20
 
-    DelayBallSpeedMultiplier = 1
+    # How "likely" it is for role assignment to choose the closest robot
+    # Higher means more likely
+    # 1 is the default number for role assignment in general
+    # Only used for collect phase
+    POSITION_COST_MULTIPLIER = 30
 
     class State(Enum):
-        intercept = 0
-        coarse_approach = 1
-        fine_approach = 2
-        delay = 3
+        # May already have the ball so just sit for a split second and keep dribbler on
+        captured = 0
+
+        # Don't have ball so try to intercept the moving ball
+        settle = 2
+
+        # Don't have ball so try to do the fine approach to control the ball
+        collect = 1
 
     ## Capture Constructor
     # faceBall - If false, any turning functions are turned off,
     # useful for using capture to reflect/bounce moving balls.
     def __init__(self, faceBall=True):
         super().__init__(continuous=False)
 
-        self.add_state(Capture.State.intercept,
-                       behavior.Behavior.State.running)
-        self.add_state(Capture.State.coarse_approach,
-                       behavior.Behavior.State.running)
-        self.add_state(Capture.State.fine_approach,
-                       behavior.Behavior.State.running)
-        self.add_state(Capture.State.delay,
-                       behavior.Behavior.State.running)
+        for s in Capture.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        self.prev_ball = main.ball()
+        self.probably_held_cnt = Capture.PROBABLY_HELD_START
+        self.frames_in_captured = 0
+
+        
+        # By default, move into the settle state since we almost never start with ball
+        self.add_transition(behavior.Behavior.State.start,
+                            Capture.State.settle,
+                            lambda: self.robot is not None and not evaluation.ball.robot_has_ball(self.robot), #self.robot.has_ball(),
```

Comment:
Until the robots get break beam sensors to detect the ball, we have to use this jank way. There's two ways we can deal with this, have both and just assume that if the break beam is valid and the ball is in the right spot, then we are good.

The only challenge is that when the ball gets covered by the robot so the cameras can't see it, it won't trigger since the robot will move the ball slightly